src/meta-srv/src/handler.rs

```diff
@@ -103,13 +114,219 @@ impl HeartbeatHandlerGroup {
+        }
+
+        ins
+    }
+
+    async fn check_timeout_bg(&self, interval_millis: u64) {
+        let mut interval = tokio::time::interval(Duration::from_millis(interval_millis));
+
+        loop {
+            interval.tick().await;
+
+            let now = Duration::from_millis(common_time::util::current_time_millis() as u64);
+            let timeout_ids = self
+                .timeouts
+                .iter()
+                .filter_map(|entry| {
+                    let (id, deadline) = entry.pair();
+                    if deadline < &now {
+                        Some(*id)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            for id in timeout_ids {
+                let _ = self
+                    .on_recv(id, Err(error::MailboxTimeoutSnafu { id }.build()))
+                    .await;
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Mailbox for HeartbeatMailbox {
+    async fn send(&self, ch: &Channel, mut msg: MailboxMessage) -> Result<MessageId> {
+        let message_id = self.sequence.next().await?;
+
+        msg.id = message_id;
+        let res = HeartbeatResponse {
+            header: None,
+            mailbox_messages: vec![msg],
+        };
+
+        let pusher_id = match ch {
+            Channel::Datanode(id) => format!("0-{}", id),
+            Channel::Frontend(id) => format!("1-{}", id),
+        };
+        let pushers = self.pushers.read().await;
+        let pusher = pushers
+            .get(&pusher_id)
+            .context(error::PusherNotFoundSnafu { pusher_id })?;
+        pusher.send(Ok(res)).await.map_err(|e| {
+            error::PushMessageSnafu {
+                err_msg: e.to_string(),
+            }
+            .build()
+        })?;
+
+        let (tx, rx) = oneshot::channel();
+        self.senders.insert(message_id, Tx(tx));
+        self.recervers.insert(message_id, Rx(rx));
+
+        Ok(message_id)
+    }
+
+    async fn recv(&self, id: MessageId) -> Result<MailboxMessage> {
+        let (_, rx) = self
+            .recervers
+            .remove(&id)
+            .context(error::MailboxNotFoundSnafu { id })?;
+        rx.0.await
+            .map_err(|_| error::MailboxClosedSnafu { id }.build())?
+    }
+
+    async fn recv_timeout(&self, id: MessageId, timeout: Duration) -> Result<MailboxMessage> {
+        let deadline =
+            Duration::from_millis(common_time::util::current_time_millis() as u64) + timeout;
+        self.timeouts.insert(id, deadline);
+        self.recv(id).await
+    }
+
+    async fn on_recv(&self, id: MessageId, maybe_msg: Result<MailboxMessage>) -> Result<()> {
+        self.timeouts.remove(&id);
+        if let Some((_, tx)) = self.senders.remove(&id) {
+            tx.0.send(maybe_msg)
+                .map_err(|_| error::MailboxClosedSnafu { id }.build())?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Can we test the 3 maps' sizes?

Reply:
Timeout checker can ensure that every `send` has a corresponding `recv`.
Except for cases where `recv` is not called, there will be no memory leak.